R/question_radio.R

```diff
@@ -53,7 +53,7 @@ question_ui_initialize.learnr_radio <- function(question, value, ...) {
     label = question$question,
     choiceNames = choice_names,
     choiceValues = choice_values,
-    selected = value %||% FALSE # setting to NULL, selects the first item
+    selected = value %||% NA # avoid selecting the first item when value is NULL
```

Comment:
If we use `NA`, it will have the same issue as `FALSE` in that the character representation will be used to match the values. (So someone with a `"NA"` choice will now be auto selected.)

Nice find about using `character(0)`.

```r
as.character(NA) %>% cat()
#> NA

as.character(character(0)) %>% cat()
#> 
```

```suggestion
    selected = value %||% character(0) # avoid selecting the first item when value is NULL
```

Reply:
Funny, I thought the same thing but in testing I couldn't get `NA` to be auto-selected. Still `character(0)` seems overall safer, so I'm happy to go with that default.

Reply:
👍 

Was trying to protect against the `as.character(selected)` in https://github.com/rstudio/shiny/blob/1558c848f4800db613703d9bc28fb9c5cdef5e14/R/input-radiobuttons.R#L97

However, `as.character(NA)` produces `NA`, not `"NA"` like `FALSE` turns into `"FALSE"`. 🤦  So `NA` will never match `"NA"`. 

But `NA` does match `NA`...

```r
shinyApp(
	ui = radioButtons(
		"id", "radio", 
		c("A", "B", "NA", NA), 
		selected = NA
	), 
	function(input, output){}
)
```

![Screen Shot 2021-04-22 at 11 13 22 AM](https://user-images.githubusercontent.com/93231/115739285-c953dd00-a35b-11eb-82a4-53a78461acc5.png)

--------------------

For some reason, a `character(0)` input gets converted to `list()` 🤷 

```r
debugonce(shiny::generateOptions)
shinyApp(
	ui = radioButtons(
		"id", "radio", 
		list(a = "A", b = "B", c = "NA", d = "FALSE", e = character(0)), 
		selected = character(0)
	), 
	function(input, output){}
)
choiceValues
#> [[1]]
#> [1] "A"
#> 
#> [[2]]
#> [1] "B"
#> 
#> [[3]]
#> [1] "NA"
#> 
#> [[4]]
#> [1] "FALSE"
#> 
#> [[5]]
#> [1] "list()"
```

Reply:
Knock knock  
_Who's there?_  
Orange... 
<details><summary><em>Orange who?</em></summary>

Orange you glad I didn't say `c("B", "A", "NA", NA)`?

</details>

Reply:
But seriously, thanks for the deep dive. Am I reading it correctly that `character(0)` is still the best choice?

Reply:
Correct. Let's use `character(0)`